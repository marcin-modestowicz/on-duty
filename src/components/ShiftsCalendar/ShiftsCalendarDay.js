// @flow
import React, { Component } from "react";
import { observer } from "mobx-react";
import type User from "../../models/User";
import styles from "./ShiftsCalendarDay.scss";

type Props = {
  day: number,
  users: User[]
};

@observer
class ShiftsCalendarDay extends Component<Props> {
  render() {
    return (
      <div className={styles.root}>
        <div className={styles.shift}>
          {this.props.users.map(({ id, name }) => (
            <div key={id} className={styles.user}>
              {name}
            </div>
          ))}
        </div>
        <div className={styles.day}>
          <div className={styles.name}>Monday</div>
          <div className={styles.number}>{this.props.day}</div>
        </div>
      </div>
    );
  }
}

export default ShiftsCalendarDay;
