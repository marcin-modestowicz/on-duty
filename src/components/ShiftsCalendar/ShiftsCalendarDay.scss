@import '../../styles/colors.scss';

.root {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  width: 10.5em;
  height: 10em;
  margin: 0 .5em .5em 0;
  box-sizing: border-box;
  border-width: 0 1px 1px 0;
  border-style: solid;
  border-radius: .25em;
  padding: .75em;
  cursor: pointer;
  user-select: none;
  text-align: center;
  background-color: $yellow;
  color: white;

  .shift {
    font-size: .8em;

    .user {
      margin-bottom: .5em;
    }
  }

  .day {
    .name {
      margin-bottom: .5em;
      font-size: .5em;
    }

    .number {
      font-size: 2em;
      font-weight: bold;

      &:after {
        content: '';
        display: block;
        margin-top: .1em;
        width: 100%;
        height: 1px;
        background-color: $yellow-darker;
        border-bottom: 1px solid $yellow-lighter;
      }
    }
  }
}
