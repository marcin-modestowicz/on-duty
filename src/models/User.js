// @flow
import { AvailabilityCalendar } from "./Calendar";
import { type AvailabilityStatus } from "./Availability";
import uuid from "uuid/v1";

export const POWER = {
  RESIDENT: 0,
  DOCTOR: 1,
  SPECIALIST: 2
};

const REQUIRED_SHIFTS = {
  RESIDENT: 2,
  DOCTOR: 1,
  SPECIALIST: 1
};

class User {
  id: string;
  name: string;
  isDoctor: boolean;
  isSpecialist: boolean;
  isAdmin: boolean;
  availabilityCalendar: AvailabilityCalendar;

  constructor(
    name: string,
    isDoctor?: boolean,
    isSpecialist?: boolean,
    availability?: AvailabilityStatus[]
  ) {
    this.id = uuid();
    this.name = name;
    this.isDoctor = !!isDoctor;
    this.isSpecialist = !!isSpecialist;
    this.isAdmin = false;
    this.availabilityCalendar = new AvailabilityCalendar(availability);
  }

  get power(): number {
    let power = POWER.RESIDENT;

    if (this.isDoctor) {
      power += POWER.DOCTOR;
    }

    if (this.isSpecialist) {
      power += POWER.SPECIALIST;
    }

    return power;
  }

  get requiredShifts(): number {
    let requiredShifts = REQUIRED_SHIFTS.RESIDENT;

    if (this.isDoctor) {
      requiredShifts = REQUIRED_SHIFTS.DOCTOR;
    }

    if (this.isSpecialist) {
      requiredShifts = REQUIRED_SHIFTS.SPECIALIST;
    }

    return requiredShifts;
  }
}

export default User;
