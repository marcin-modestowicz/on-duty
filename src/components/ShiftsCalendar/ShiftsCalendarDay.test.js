// @flow
import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import User from "../../models/User";
import ShiftsCalendarDay from "./ShiftsCalendarDay";

describe("ShiftsCalendarDay component", () => {
  let shiftsCalendarDay;

  beforeAll(() => {
    const users = [
      new User("user1", false, false),
      new User("user2", true, false)
    ];
    users.forEach((user, index) => {
      user.id = `a${index}`;
    });
    shiftsCalendarDay = shallow(<ShiftsCalendarDay day={1} users={users} />);
  });

  test("should render", () => {
    expect(toJson(shiftsCalendarDay)).toMatchSnapshot();
  });
});
