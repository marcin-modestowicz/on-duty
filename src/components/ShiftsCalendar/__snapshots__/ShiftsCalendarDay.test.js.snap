// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShiftsCalendarDay component should render 1`] = `
<div
  className="root"
  onClick={[Function]}
>
  <div
    className="shift"
  >
    <div
      className="user"
      key="a0"
    >
      user1
    </div>
    <div
      className="user"
      key="a1"
    >
      user2
    </div>
  </div>
  <div
    className="day"
  >
    <div
      className="name"
    >
      Sunday
    </div>
    <div
      className="number"
    >
      1
    </div>
  </div>
</div>
`;
